### PR TITLE
chore(networkd): Report on errors during interface configuration

### PR DESCRIPTION
**Pull request by bradbeam**
This DRYs up the interface configuration and adds in an error channel to capture
any issues that come up from interface configuration. These errors are still
treated as non-fatal, but should provide some additional insight.

Signed-off-by: Brad Beam <brad.beam@b-rad.info>

### CODE REVIEW

File: internal/app/networkd/pkg/networkd/networkd.go [hunk 1]
```diff
@@ -410,3 +348,52 @@ func (n *Networkd) SetReady() {
...
+		if iface.Bonded != bonded {
+			continue
+		}
+
+		wg.Add(1)
+
+		go func(netif *nic.NetworkInterface, wg *sync.WaitGroup) {
+			defer wg.Done()
+
+			log.Printf("setting up %s", netif.Name)
```
Comment (smira) [inline internal/app/networkd/pkg/networkd/networkd.go]: one potential way to DRY up this function a bit is to do something like:

```go
errCh <- func() error {
   			if err := netif.Create(); err != nil {
				return fmt.Errorf("error creating nic: %v", err)
			}
                        
                        if err = ...
        }
}
```

File: internal/app/networkd/pkg/networkd/networkd.go [hunk 2]
```diff
@@ -410,3 +348,52 @@ func (n *Networkd) SetReady() {
...
+
+			// Ensure link exists
+			if err = netif.Create(); err != nil {
```
Comment (smira) [inline internal/app/networkd/pkg/networkd/networkd.go]: this write is a data race, variable `err` is used from multiple goroutines

should be `if err := netif.Create()`

File: internal/app/networkd/pkg/networkd/networkd.go [hunk 3]
```diff
@@ -410,3 +348,52 @@ func (n *Networkd) SetReady() {
...
+	)
+
+	for _, iface := range n.Interfaces {
+		if iface.Bonded != bonded {
+			continue
+		}
+
+		wg.Add(1)
+
+		go func(netif *nic.NetworkInterface, wg *sync.WaitGroup) {
```
Comment (smira) [inline internal/app/networkd/pkg/networkd/networkd.go]: usually `wg` is not passed to the closure, as it's not a for loop variable, implicit variable access from outer scope is just fine

File: internal/app/networkd/pkg/networkd/networkd.go [hunk 4]
```diff
@@ -160,78 +160,16 @@ func New(config runtime.Configurator) (*Networkd, error) {
 // configuration, and any addressing that is needed. We care about ordering
 // here so that we can ensure any links that make up a bond will be in
 // the correct state when we get to bonding configuration.
-// nolint: gocyclo
 func (n *Networkd) Configure() (err error) {
-	var wg sync.WaitGroup
-
-	log.Println("configuring non-bonded interfaces")
```
Comment (smira) [inline internal/app/networkd/pkg/networkd/networkd.go]: should we keep those log message so it's easy to see what's going on?

File: internal/app/networkd/pkg/networkd/networkd.go [hunk 5]
```diff
@@ -410,3 +356,45 @@ func (n *Networkd) SetReady() {
...
+
+		count++
+
+		go func(netif *nic.NetworkInterface) {
+			log.Printf("setting up %s", netif.Name)
+
+			errCh <- func() error {
+				// Ensure link exists
+				if err := netif.Create(); err != nil {
+					return fmt.Errorf("error creating nic: %v", err)
```
Comment (smira) [inline internal/app/networkd/pkg/networkd/networkd.go]: I guess we need to add NIC name to the error message to make it easy to match error & config

Comment (andrewrynhard) [reply internal/app/networkd/pkg/networkd/networkd.go]: Good point

**Pull request merged**